In [9]:
import requests
import pandas as pd
from datetime import datetime, date, timedelta
import numpy as np
import re
import math

In [10]:
# гугл монтер, надо будет вводить AUTH код для этого
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [11]:
# датасет где-то потерялся, поэтому пришлось перепарсить используя 3ю лабу
df = pd.read_csv('/content/gdrive/My Drive/ABD_Lab_dataset/hh_parsed.csv', delimiter=',')
df = df.drop(columns=['Unnamed: 0'])
df

name  ...                                        requirement
0      Системный архитектор НИОКР (вычислительная тех...  ...  Высшее техническое образование. Опыт практичес...
1                                 Java Developer Trainee  ...  Уверенное знание принципов ООП, алгоритмов и с...
2                                 Java Developer Trainee  ...  Уверенное знание принципов ООП, алгоритмов и с...
3                              Начальник отдела развития  ...  Высшее финансовое / техническое / юридическое ...
4         Главный инженер контроля качества тестирования  ...  Опыт работы в области IT (QA/Разработка/Тестир...
...                                                  ...  ...                                                ...
48040                                      Монтажник ЛКС  ...  опыт в сфере телекоммуникационных услуг и энер...
48041                                     Программист JS  ...  Хорошее знание JavaScript, html, css. Опыт раб...
48042                                 Программист Python  ...  Знание Python или иных скриптовых языков с пос...
48043                                  Оператор ЭВМ (ПК)  ...  Отличное владение ПК, знание офисных программ ...
48044                              Менеджер отдела ЕГАИС  ...  Высшее техническое образование (будет являться...

[48045 rows x 9 columns]

In [12]:
# все уникальные типы занятости (нужный для одного из заданий)
types = df['schedule'].value_counts().keys() 
types

Index(['Полный день', 'Удаленная работа', 'Гибкий график', 'Сменный график',
       'Вахтовый метод'],
      dtype='object')

In [13]:
unique_names = df['name'].value_counts().keys() 
unique_names

Index(['Системный администратор', 'Программист 1С', 'Менеджер по продажам',
       'Оператор саll-центра', 'Специалист технической поддержки',
       'Системный аналитик',
       'Специалист по продаже и установке спутникового ТВ', 'Программист 1C',
       'Frontend-разработчик', 'Менеджер по продажам В2В для Яндекс.Директа',
       ...
       'Руководитель проектов по ИТ-инфраструктуре',
       'Инженер-программист (цифровая обработка сигналов)',
       'Стажер отдела маркетинга',
       'Консультант 1С ( Тверская область, г. Конаково)',
       'Ведущий бизнес-аналитик (ecommerce)',
       'Наставник аналитиков в IT-проекте', 'Product marketing analyst',
       'Ведущий аналитик (МТС ТВ)', 'Менеджер (ИТ-Партнер)',
       'Специалист слаботочных систем'],
      dtype='object', length=24159)



---


Задание: придумать подход и сгруппировать вакансии по названиям (например, Web-разработчик и Web-Developer можно объединить в одну группу);

Исходя из блока кода выше `unique_names` содержит в себе всевозможные варианты
того какие профессии есть (их 24159 штук). 

Для группировки можно использовать несколько подходов:
1.   Группировать по "базовым" словам. Например: `программист`, `администратор` и т.д.
2.   Так как названия могут и не содержать "базовых" слов из пункта 1. То можно сплитать по словам и на их основе делать группы. Проблема данного алгоритма в том, что он может оказаться достаточно турднозатратным по времени + может вызвать небольшую сложность в его написании и дебаге.
3.   Подрубить обученную нейронку и она сама все сделает (такой нейронки не нашел)

Вследствие этого будем использовать подход **1**



---



In [14]:
# норм работает
#df.name.str.lower().str.find('программист') == 0
#df.loc[df.name.str.lower().str.find('программист') == 0, 'name'] = 'программист'
#df.loc[df.name == 'программист']

# массив выделенных мной "базовых слов"
based_words = ['Developer','Разработчик' ,'Программист', 'Специалист', 'Инженер', 'Оператор', 'Администратор', 'Аналитик']
for word in based_words:
  if word.lower() == 'developer':
    df.loc[df.name.str.lower().str.find(word.lower()) == 0, 'category'] = 'Разработчик'
  else:
    df.loc[df.name.str.lower().str.find(word.lower()) == 0, 'category'] = word
# то что не подошло просто ставим другое
df.category.fillna('Другое', inplace=True)
df

name  ...     category
0      Системный архитектор НИОКР (вычислительная тех...  ...       Другое
1                                 Java Developer Trainee  ...       Другое
2                                 Java Developer Trainee  ...       Другое
3                              Начальник отдела развития  ...       Другое
4         Главный инженер контроля качества тестирования  ...       Другое
...                                                  ...  ...          ...
48040                                      Монтажник ЛКС  ...       Другое
48041                                     Программист JS  ...  Программист
48042                                 Программист Python  ...  Программист
48043                                  Оператор ЭВМ (ПК)  ...     Оператор
48044                              Менеджер отдела ЕГАИС  ...       Другое

[48045 rows x 10 columns]



---

Задание: добавить новый признак в датасет на основе признака “дата размещения вакансии” - количество дней с момента размещения;

---



In [15]:
# Если делать так то будет OutOfMemory для VM, почему так происходит понять я не могу
# поэтому придется использовать циклы, так как и на подобных манипуляциях происходит out of memory

#days_series = [round(abs(datetime.now() - datetime.strptime(x[:10], "%Y-%m-%d")).days) for x in df.published_at]
#ser = pd.Series(days_series)
#df = pd.concat([df, ser.to_frame().T])
#df

In [16]:
for a in range(len(df)):
  if pd.isna(df.loc[a, 'minSalary']):
    sorted = df.loc[(df['area'] == df.loc[a, 'area']) & (df['category'] == df.loc[a, 'category'])]
    mean = sorted['minSalary'].mean()
    # средняя з/п тоже может быть NaN
    if not math.isnan(mean): 
      df.loc[a, 'minSalary'] = round(mean)
    
  if pd.isna(df.loc[a, 'maxSalary']):
    sorted = df.loc[(df['area'] == df.loc[a, 'area']) & (df['category'] == df.loc[a, 'category'])]
    mean = sorted['maxSalary'].mean()
    if not math.isnan(mean):
      df.loc[a, 'maxSalary'] = round(mean)

  # - добавить новый признак в датасет на основе признака “дата размещения вакансии” - количество дней с момента размещения;
  df.loc[a, 'days'] = round(abs(datetime.now()-datetime.strptime(df.loc[a, 'published_at'][:10], "%Y-%m-%d")).days)
  
  # - заполнить пропуски в признаке “требуемый опыт работы” по принципу: если не указан, то опыт не требуется (нет в датасете);
  
  # - заполнить пропуски в признаке “тип занятости” по принципу: если не указан, то любой тип;
  if pd.isna(df.loc[a, 'schedule']):
    df.loc[a, 'schedule'] = random.choice(types)

  # - при наличии пропусков в остальных признаках заполнить их по методу на свое усмотрение.
  if pd.isna(df.loc[a, 'responsibility']):
    df.loc[a, 'responsibility'] = 'Отсутствуют'
  if pd.isna(df.loc[a, 'requirement']):
    df.loc[a, 'requirement'] = 'Отсутствуют'

df

name  ...  days
0      Системный архитектор НИОКР (вычислительная тех...  ...  30.0
1                                 Java Developer Trainee  ...  30.0
2                                 Java Developer Trainee  ...  30.0
3                              Начальник отдела развития  ...  30.0
4         Главный инженер контроля качества тестирования  ...  30.0
...                                                  ...  ...   ...
48040                                      Монтажник ЛКС  ...   0.0
48041                                     Программист JS  ...   0.0
48042                                 Программист Python  ...   0.0
48043                                  Оператор ЭВМ (ПК)  ...   0.0
48044                              Менеджер отдела ЕГАИС  ...   0.0

[48045 rows x 11 columns]